src/Native/FactorizationMachineNative/FactorizationMachineCoreAVX.cpp

```diff
@@ -0,0 +1,178 @@
+            float * qffprime = pq + qBias;
+
+            // q_f,f' += v_j,f' * x
+            for (int k = 0; k + 8 <= d; k += 8)
+            {
+                const __m256 _v = _mm256_load_ps(vjfprime + k);
+                __m256 _q = _mm256_load_ps(qffprime + k);
+                _q = _mm256_add_ps(_q, _mm256_mul_ps(_v, _x));
+                _mm256_store_ps(qffprime + k, _q);
+            }
+        }
+    }
+
+    for (int f = 0; f < m; f++)
+    {
+        // tmp += <q_f,f, q_f,f>
+        const float * qff = pq + f * m * d + f * d;
+        for (int k = 0; k + 8 <= d; k += 8)
+        {
+            __m256 _qff = _mm256_load_ps(qff + k);
+
+            // Intra-field interactions.
+            _tmp = _mm256_add_ps(_tmp, _mm256_mul_ps(_qff, _qff));
+        }
+
+        // y += <q_f,f', q_f',f>, f != f'
+        // Whis loop handles inter - field interactions because f != f'.
+        for (int fprime = f + 1; fprime < m; fprime++)
+        {
+            const float * qffprime = pq + f * m * d + fprime * d;
+            const float * qfprimef = pq + fprime * m * d + f * d;
+            for (int k = 0; k + 8 <= d; k += 8)
+            {
+                // Inter-field interaction.
+                __m256 _qffprime = _mm256_load_ps(qffprime + k);
+                __m256 _qfprimef = _mm256_load_ps(qfprimef + k);
+                _y = _mm256_add_ps(_y, _mm256_mul_ps(_qffprime, _qfprimef));
+            }
+        }
+    }
+
+    _y = _mm256_add_ps(_y, _mm256_mul_ps(_mm256_set1_ps(0.5f), _tmp));
+    _tmp = _mm256_add_ps(_y, _mm256_permute2f128_ps(_y, _y, 1));
+    _tmp = _mm256_hadd_ps(_tmp, _tmp);
+    _y = _mm256_hadd_ps(_tmp, _tmp);
+    _mm_store_ss(&latentResponse, _mm256_castps256_ps128(_y));
+    *response = linearResponse + latentResponse;
+
+}
+
+// Calculate the stochastic gradient and update the model.
+// The /*const*/ comment on the parameters of the function means that their values should not get altered by this function.
+EXPORT_API(void) CalculateGradientAndUpdateNativeAVX(float lambdaLinear, float lambdaLatent, float learningRate, int fieldCount, int latentDim, float weight, int count,
+    _In_ int* /*const*/ fieldIndices, _In_ int* /*const*/ featureIndices, _In_ float* /*const*/ featureValues, _In_ float* /*const*/ latentSum, float slope,
+    _Inout_ float* linearWeights, _Inout_ float* latentWeights, _Inout_ float* linearAccumulatedSquaredGrads, _Inout_ float* latentAccumulatedSquaredGrads)
+{
+    const int m = fieldCount;
+    const int d = latentDim;
+    const int c = count;
+    const int * pf = fieldIndices;
+    const int * pi = featureIndices;
+    const float * px = featureValues;
+    const float * pq = latentSum;
+    float * pw = linearWeights;
+    float * pv = latentWeights;
+    float * phw = linearAccumulatedSquaredGrads;
+    float * phv = latentAccumulatedSquaredGrads;
+
+    const __m256 _wei = _mm256_set1_ps(weight);
+    const __m256 _s = _mm256_set1_ps(slope);
```

Comment:
Does our memory alignment meet AVX's requirement?

Reply:
Thanks for pointing it out. I will take a look at the memory alignment for AVX/FMA.